rust/parquet/src/record/reader.rs

```diff
@@ -15,687 +15,1124 @@
-    /// Creates new tree builder with default parameters.
-    pub fn new() -> Self {
-        Self {
-            batch_size: DEFAULT_BATCH_SIZE,
+impl<A, B, C> Reader for sum::Sum3<A, B, C>
+where
+    A: Reader,
+    B: Reader<Item = A::Item>,
+    C: Reader<Item = A::Item>,
+{
+    type Item = A::Item;
+
+    #[inline]
+    fn read(&mut self, def_level: i16, rep_level: i16) -> Result<Self::Item> {
+        match self {
+            sum::Sum3::A(ref mut reader) => reader.read(def_level, rep_level),
+            sum::Sum3::B(ref mut reader) => reader.read(def_level, rep_level),
+            sum::Sum3::C(ref mut reader) => reader.read(def_level, rep_level),
         }
     }
 
-    /// Sets batch size for this tree builder.
-    pub fn with_batch_size(mut self, batch_size: usize) -> Self {
-        self.batch_size = batch_size;
-        self
+    #[inline]
+    fn advance_columns(&mut self) -> Result<()> {
+        match self {
+            sum::Sum3::A(ref mut reader) => reader.advance_columns(),
+            sum::Sum3::B(ref mut reader) => reader.advance_columns(),
+            sum::Sum3::C(ref mut reader) => reader.advance_columns(),
+        }
     }
 
-    /// Creates new root reader for provided schema and row group.
-    pub fn build(
-        &self,
-        descr: SchemaDescPtr,
-        row_group_reader: &RowGroupReader,
-    ) -> Reader {
-        // Prepare lookup table of column path -> original column index
-        // This allows to prune columns and map schema leaf nodes to the column readers
-        let mut paths: HashMap<ColumnPath, usize> = HashMap::new();
-        let row_group_metadata = row_group_reader.metadata();
+    #[inline]
+    fn has_next(&self) -> bool {
+        match self {
+            sum::Sum3::A(ref reader) => reader.has_next(),
+            sum::Sum3::B(ref reader) => reader.has_next(),
+            sum::Sum3::C(ref reader) => reader.has_next(),
+        }
+    }
 
-        for col_index in 0..row_group_reader.num_columns() {
-            let col_meta = row_group_metadata.column(col_index);
-            let col_path = col_meta.column_path().clone();
-            paths.insert(col_path, col_index);
+    #[inline]
+    fn current_def_level(&self) -> i16 {
+        match self {
+            sum::Sum3::A(ref reader) => reader.current_def_level(),
+            sum::Sum3::B(ref reader) => reader.current_def_level(),
+            sum::Sum3::C(ref reader) => reader.current_def_level(),
         }
+    }
 
-        // Build child readers for the message type
-        let mut readers = Vec::new();
-        let mut path = Vec::new();
+    #[inline]
+    fn current_rep_level(&self) -> i16 {
+        match self {
+            sum::Sum3::A(ref reader) => reader.current_rep_level(),
+            sum::Sum3::B(ref reader) => reader.current_rep_level(),
+            sum::Sum3::C(ref reader) => reader.current_rep_level(),
+        }
+    }
+}
+
+// ----------------------------------------------------------------------
```

Comment:
Please use macro + specialization to reduce redundant code in the following.

Reply:
Sure thing